### PR TITLE
Reuse PermanentNavModel if provided in the list of navModels

### PR DESCRIPTION
**Pull request by KovalevAndrey**
## Description

Fixes: https://github.com/bumble-tech/appyx/issues/604

Problem: If `PermanentNavModel` was provided to the constructor of the ParentNode it will be ignored as ParenNode internally uses its own instance of `PermanentNavModel`. This PR ensures that we reuse the same instance of `PermanentNavModel` if it's provided 

## Check list

- [x] I have updated `CHANGELOG.md` if required.
- [x] I have updated documentation if required.

### CODE REVIEW

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt [hunk 1]
```diff
@@ -62,11 +63,52 @@ abstract class ParentNode<NavTarget : Any>(
...
+            navModel + permanentNavModel
+        }
+    }
+
+    private fun retrievePermanentNavModel(navModel: NavModel<NavTarget, *>): PermanentNavModel<NavTarget>? =
+        when (navModel) {
+            is CombinedNavModel<NavTarget> -> {
+                navModel
+                    .navModels
+                    .find { it is PermanentNavModel<NavTarget> } as PermanentNavModel<NavTarget>?
```
Comment (LachlanMcKee) [inline libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:97]: It seems like it's possible (though maybe not advised?) that a developer could nest a `CombinedNavModel` within a `CombinedNavModel`.

Perhaps this case should be covered as well? (if so, then the init block for the CombinedNavModel might need to be updated).

If we don't want to allow nesting `CombinedNavModel` within `CombinedNavModel` perhaps the `init` block of `CombinedNavModel` could prevent this as well? (though this hypothetically could be a breaking change.

i.e.

```
CombinedNavModel(
    listOf(
         CombinedNavModel(
             listOf(
                 PermanentNavModel()
             )
         )
    )
)
```

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/model/combined/CombinedNavModel.kt [hunk 2]
```diff
@@ -18,6 +19,13 @@ class CombinedNavModel<NavTarget>(
     val navModels: List<NavModel<NavTarget, *>>,
 ) : NavModel<NavTarget, Any?>, Destroyable {
 
+    init {
+        val permanentNavModelCount = navModels.filterIsInstance<PermanentNavModel<*>>().count()
```
Comment (LachlanMcKee) [inline libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/model/combined/CombinedNavModel.kt]: Slightly more efficient (as it doesn't create, then discard a list):
`val permanentNavModelCount = navModels.count { it is PermanentNavModel<*> }`

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/model/combined/CombinedNavModel.kt [hunk 3]
```diff
@@ -18,6 +19,13 @@ class CombinedNavModel<NavTarget>(
...
+        check(permanentNavModelCount <= MAX_PERMANENT_NAV_MODEL_COUNT) {
+            "Do not provide more than one PermanentNavModel"
```
Comment (LachlanMcKee) [inline libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/model/combined/CombinedNavModel.kt]: maybe reword to: `CombinedNavModel does not support more than one PermanentNavModel`?

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt [hunk 4]
```diff
@@ -62,11 +63,52 @@ abstract class ParentNode<NavTarget : Any>(
     plugins = plugins + navModel + childAware
 ), Resolver<NavTarget> {
 
-    private val permanentNavModel = PermanentNavModel<NavTarget>(
-        savedStateMap = buildContext.savedStateMap,
-        key = KEY_PERMANENT_NAV_MODEL,
-    )
-    val navModel: NavModel<NavTarget, *> = permanentNavModel + navModel
+    private lateinit var permanentNavModel: PermanentNavModel<NavTarget>
+    private var isPermanentNavModelProvided: Boolean? = null
```
Comment (LachlanMcKee) [inline libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt]: Does this need to be nullable? perhaps it could be `lateinit` as well?

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt [hunk 5]
```diff
@@ -62,11 +63,52 @@ abstract class ParentNode<NavTarget : Any>(
...
+            }
+            is PermanentNavModel<NavTarget> -> navModel
+            else -> null
+        }
+
+    private fun createPermanentNavModel(buildContext: BuildContext): PermanentNavModel<NavTarget> =
+        PermanentNavModel(
+            savedStateMap = buildContext.savedStateMap,
+            key = KEY_PERMANENT_NAV_MODEL,
+        )
```
Comment (LachlanMcKee) [inline libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt]: This seems to be unused. was the intention to call this from `resolveNavModels`?

Comment (KovalevAndrey) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt]: yeah, detekt already complained

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt (hunk 4 above)
Comment (KovalevAndrey) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt]: primitives can't be used with lateinit

Comment (CherryPerry) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt]: Let's drop `lateinit` instead?

```kotlin
val navModel: NavModel<NavTarget, *>
private val permanentNavModel: PermanentNavModel<NavTarget>
private val isPermanentNavModelProvided: Boolean

init {
   val (combinedNavModel, permanentNavModel, isProvided) = resolveNavModels(navModel, buildContext)
   this.navModel = combinedNavModel
   this.permanentNavModel = permanentNavModel
   isPermanentNavModelProvided = isProvided
}
```

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt (hunk 1 above)
Comment (KovalevAndrey) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:97]: yeah, I think we can prevent nesting of `CombinedNavModel` as it doesn't make much sense and harder to support

**Review by CherryPerry — COMMENTED** [review]
In general OK, but IMHO we are fixing the wrong thing. The library is `Model-driven navigation`, but `@Composable fun PermanentChild(key)` does not fit really into this category, making "navigation" UI-driven. I think we just should provide more convenient `BagNavModel` (that we already use in tests) where we can freely use `add` and `remove` operations and have a Composable function to render its children one by one (by key as it is right now). We should not have such weird shortcuts for child hosting as we have right now and then write a lot of code around handling them.

File: libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt (hunk 1 above)
Comment (CherryPerry) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:97]: Nesting makes sense at the current moment.
First you pass `CombinedNavModel` via `ParentNode(navModel = backStack1 + backStack2)` then this `CombinedNavModel` is merged with internal `PermanentNavModel` into another `CombinedNavModel`. So at the end we have `CombinedNavModel` inside `CombinedNavModel` is a pretty basic case.

Comment (LachlanMcKee) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:97]: @KovalevAndrey it sounds like this might be wise to keep? otherwise you should definitely update the changelog to call this out as a breaking change.

Comment (KovalevAndrey) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:97]: @LachlanMcKee Looking at the code I'm leaning towards removing the `PermanentNavModel` from parentNode at all and use it the same way as any other navModels

Comment (KovalevAndrey) [reply libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:97]: > Nesting makes sense at the current moment. First you pass `CombinedNavModel` via `ParentNode(navModel = backStack1 + backStack2)` then this `CombinedNavModel` is merged with internal `PermanentNavModel` into another `CombinedNavModel`. So at the end we have `CombinedNavModel` inside `CombinedNavModel` is a pretty basic case.

If we're using the provided API it doesn't get nested

Screenshot 2023-09-30 at 10 48 19
